File: Chome_Exten/ProxyServer/ProxyServer.js
const net = require('net');
const { hostname } = require('os');
const WebSocket = require('ws');
const app = net.createServer();


const PORT = process.env.PORT || 8080;
var forbiddenSites = {};



const UnSafeSite = "127.0.0.1:8089";

app.on("connection", (clientToProxySocket) => {
  console.log("Client connected to Proxy");

  clientToProxySocket.once("data", async (data) => {
    console.log(data.toString());
    let isConnectionTLS = data.toString().indexOf("CONNECT") !== -1;
    console.log(isConnectionTLS.toString());

    let serverPort = 80;
    let serverAddr;

    if (isConnectionTLS) {
      serverPort = 443;
      serverAddr = data.toString().split("CONNECT")[1].split(" ")[1].split(":")[0];
      console.log(serverAddr);
    } else {
      serverAddr = data.toString().split("Host: ")[1].split("\n")[0].trim();
    }

    const options = {
      host: '127.0.0.1',
      port: 8069,
      //ca: [fs.readFileSync('path/to/ca-cert.pem')],
      secureProtocol: 'TLSv1_2_method', // Use TLS 1.2
      ciphers: 'ECDHE-RSA-AES256-GCM-SHA384:DHE-RSA-AES256-GCM-SHA384:HIGH:!aNULL:!eNULL:!EXPORT:!DES:!RC4:!MD5:!PSK:!SRP:!CAMELLIA'
    };

    try {
      if(!(serverAddr in forbiddenSites)){
        const isAllowed = await checkSafeNet(options, serverAddr);
        if (!isAllowed) {
          serveBlackPage(clientToProxySocket);
          console.log("redirecting");
          return;
      }}
      else if(forbiddenSites[serverAddr] == 0) {
          serveBlackPage(clientToProxySocket);
          console.log("redirecting");
          return;
      }

      let proxyToServerSocket = net.createConnection({
        host: serverAddr,
        port: serverPort,
      }, () => {
        console.log("Proxy connected to server");
      });

      if (isConnectionTLS) {
        clientToProxySocket.write("HTTP/1.1 200 OK\r\n\r\n");
        console.log("Wrote back");
      }

      clientToProxySocket.pipe(proxyToServerSocket);
      proxyToServerSocket.pipe(clientToProxySocket);

      proxyToServerSocket.on("error", (err) => {
        console.log("Proxy to server error");
        console.log(err);
      });

      clientToProxySocket.on("error", (err) => {
        console.log("Client to proxy error");
      });

      proxyToServerSocket.on("close", () => {
        console.log("Server connection closed");
      });

      clientToProxySocket.on("close", () => {
        console.log("Client connection closed");
      });

    } catch (error) {
      console.error('Error connecting to SafeNet:', error);
      clientToProxySocket.end('HTTP/1.1 500 Internal Server Error\r\n\r\n');
    }
  });
});

function serveBlackPage(clientSocket) {
  const htmlContent = `
    HTTP/1.1 200 OK\r\n
    Content-Type: text/html\r\n
    \r\n
    <html>
    <head>
      <meta charset="UTF-8">
      <meta name="identifier" content="Safe-Net-Black-Page">
      <title>Access Forbidden</title>
      <style>
        body {
          background-color: black;
          color: white;
          text-align: center;
          padding-top: 50px;
        }
      </style>
    </head>
    <body>
      <h1>Access Forbidden</h1>
      <p>You are being redirected...</p>
    </body>
    </html>
  `;
clientSocket.end(htmlContent);
}
function checkSafeNet(options, serverAddr) {
  return new Promise((resolve, reject) => {
    const proxyToSafeNetSocket = net.connect(options, () => {
      const connectRequest = `CONNECT Host: ${serverAddr}\r\n\r\n`;
      console.log("-------------");
      console.log(connectRequest);
      const connectRequestBuffer = Buffer.from(connectRequest, 'utf-8');
      proxyToSafeNetSocket.write(connectRequestBuffer);
      console.log("Proxy connected to SafeNet");
    });

    proxyToSafeNetSocket.on('data', (data) => {
      const dataStr = data.toString();
      if (dataStr.includes("HTTP/1.1 203")) {
        forbiddenSites[serverAddr] = 0; // Add site to forbiddenSites with value 0
        console.log("Site forbidden");
        console.log(forbiddenSites);
        resolve(false); // Not allowed
      } else {
        resolve(true); // Allowed
      }
      proxyToSafeNetSocket.end();
    });

    proxyToSafeNetSocket.on("error", (err) => {
      console.log("Proxy to SafeNet error");
      console.log(err);
      reject(err);
    });
  });
}

app.listen({ host: '127.0.0.1', port: PORT }, () => {
  console.log("Server running on PORT:", PORT);
});


// Create a new WebSocket server
const wss = new WebSocket.Server({ port: 8089 });

// Event listener for new connections
wss.on('connection', function connection(ws) {
  console.log('extention connected!');

  // Event listener for messages from clients
  ws.on('message', function incoming(message) {
    console.log('Received message:', message);
    let hostname = message.toString().split("Host: ")[1].split("\n")[0].trim()
    if (Object.keys(forbiddenSites).includes(hostname)) {
      forbiddenSites[hostname] = 1;
      console.log(forbiddenSites);
    }
    console.log("tring to unlock V");
    console.log(hostname);

  });

  // Event listener for when a client disconnects
  ws.on('close', function close() {
    console.log('A client disconnected');
  });
});

console.log('WebSocket server is listening on ws://localhost:8081');

File: Chome_Exten/ProxyServer/ProxyServer.py
import socket
import ssl
import os
import threading

PORT = 8080
forbidden_sites = {}

def handle_client(client_socket):
    print("Client connected to Proxy")
    
    data = client_socket.recv(4096)
    print(data.decode())
    is_connection_tls = b"CONNECT" in data
    print(str(is_connection_tls))
    
    server_port = 80
    server_addr = None

    if is_connection_tls:
        server_port = 80
        server_addr = data.split(b"CONNECT")[1].split(b" ")[1].split(b":")[0].decode()
        print(server_addr)
    else:
        server_addr = data.split(b"Host: ")[1].split(b"\\n")[0].decode()

    context = ssl.create_default_context()
    context.check_hostname = False
    context.verify_mode = ssl.CERT_NONE
    
    with socket.create_connection(('127.0.0.1', 8069)) as proxy_to_safe_net_socket:
        with context.wrap_socket(proxy_to_safe_net_socket, server_hostname='127.0.0.1') as ssock:
            #connect_request = f'CONNECT {server_addr}:{server_port} HTTP/1.1\r\nHost: {server_addr}:{server_port}\r\n\r\n'.encode()
            connect_request = data
            print("-------------")
            print(connect_request.decode())
            ssock.sendall(connect_request)
            print("Proxy connected to SafeNet")
        
            if is_connection_tls:
                client_socket.sendall(b"HTTP/1.1 200 OK\r\n\r\n")
                print("Wrote back")
            else:
                if server_addr in forbidden_sites and forbidden_sites[server_addr] == 1:
                    data = replace_port(data.decode(), -1).encode()
                    print(data.decode())
                ssock.sendall(data)
            
            while True:
                try:
                    proxy_data = b''
                    while True:
                        data = ssock.recv(4096)
                        if len(data) == 0:
                            break
                        proxy_data += data
                    if not proxy_data:
                        break
                    if b" 203 " in proxy_data:
                        forbidden_sites[server_addr] = 0  # 0 is not approved, 1 is approved
                    print('Received: ' + proxy_data.decode())
                    client_socket.sendall(proxy_data)
                except Exception as e:
                    print("Proxy to server error")
                    print(e)
                    break

def replace_port(url_string, new_port):
    import re
    return re.sub(r":\d+", f":{new_port}", url_string)

def main():
    server = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    server.bind(('127.0.0.1', PORT))
    server.listen()
    print(f"Server running on PORT: {PORT}")

    try:
        while True:
            client_socket, addr = server.accept()
            threading.Thread(target=handle_client, args=(client_socket,)).start()
    except KeyboardInterrupt:
        print("Server shutting down")
    finally:
        server.close()

if __name__ == "__main__":
    main()
